File: backend/src/main/java/com/uniasselvi/lwdstore/services/CategoryService.java
package com.uniasselvi.lwdstore.services;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

import com.uniasselvi.lwdstore.repositories.CategoryRepository;
import com.uniasselvi.lwdstore.services.exceptions.DatabaseException;
import com.uniasselvi.lwdstore.services.exceptions.ResourceNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.DataIntegrityViolationException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Propagation;
import org.springframework.transaction.annotation.Transactional;

import com.uniasselvi.lwdstore.dto.CategoryDTO;
import com.uniasselvi.lwdstore.entities.Category;

import jakarta.persistence.EntityNotFoundException;

@Service
public class CategoryService {
	
	@Autowired
	private CategoryRepository categoryRepository;
	
	@Transactional(readOnly = true)
	public Page<CategoryDTO> findAllPagedByName(String name, Pageable pageable) {
		Page<Category> categories = categoryRepository.searchAllPagedByName(name, pageable);
		return categories.map(cat -> new CategoryDTO(cat));
	}
	
	@Transactional(readOnly = true)
	public List<CategoryDTO> findAll() {
		List<Category> list = categoryRepository.findAll();
		return list.stream().map(item -> new CategoryDTO(item)).collect(Collectors.toList());
	}
	
	@Transactional(readOnly = true)
	public CategoryDTO findById(Long id) {
		Optional<Category> opt = categoryRepository.findById(id);
		Category cat = opt.orElseThrow(() -> new ResourceNotFoundException("Entity not found"));
		return new CategoryDTO(cat);
	}
	
	@Transactional
	public CategoryDTO insert(CategoryDTO dto) {
		Category category = new Category();
		category.setName(dto.getName());
		category.setImageUrl(dto.getImageUrl());
		category = categoryRepository.save(category);
		return new CategoryDTO(category);
	}
	
	@Transactional
	public CategoryDTO update(Long id, CategoryDTO dto) {
		try {
			Category category = categoryRepository.getReferenceById(id);
			category.setName(dto.getName());
			category.setImageUrl(dto.getImageUrl());
			category = categoryRepository.save(category);
			return new CategoryDTO(category);
		}
		catch (EntityNotFoundException e) {
			throw new ResourceNotFoundException("Id not found " + id);
		}
	}
	
	@Transactional(propagation = Propagation.SUPPORTS)
	public void delete(Long id) {
		if(!categoryRepository.existsById(id)) {
			throw new ResourceNotFoundException("Id not found " + id);
		}
		try {
			categoryRepository.deleteById(id);
		}
		catch (DataIntegrityViolationException e) {
			throw new DatabaseException("Integrity violation");
		}
	}
}

File: frontend/src/services/category-service.ts
import { AxiosRequestConfig } from "axios";
import { requestBackend } from "../utils/requests";
import { Category } from "../types/category";

export function findAllPageable(name: string, page: number, size = 3) {
  const config: AxiosRequestConfig = {
    method: "GET",
    url: "/categories",
    params: {
      name: name,
      page: page,
      size: size,
    },
  };
  return requestBackend(config);
}

export function findAll() {
  const config: AxiosRequestConfig = {
    method: "GET",
    url: "/categories/list",
  };
  return requestBackend(config);
}

export function findById(id: number) {
  const config: AxiosRequestConfig = {
    method: "GET",
    url: `/categories/${id}`,
  };

  return requestBackend(config);
}

export function insert(category: Category) {
  const config: AxiosRequestConfig = {
    method: "POST",
    url: "categories",
    data: category,
    withCredentials: true,
  };
  return requestBackend(config);
}

export function update(id: number, category: Category) {
  const config: AxiosRequestConfig = {
    method: "PUT",
    url: `/categories/${id}`,
    data: category,
    withCredentials: true,
  };
  return requestBackend(config);
}

export function remove(id: number) {
  const config: AxiosRequestConfig = {
    method: "DELETE",
    url: `/categories/${id}`,
    withCredentials: true,
  };
  return requestBackend(config);
}
